File: backend/src/main/kotlin/no/kartverket/geonorge/kartkatalog/metadata/DistributionProtocols.kt
package no.kartverket.geonorge.kartkatalog.metadata

object DistributionProtocols {
    private val VIEW_SERVICES =
        setOf(
            "OGC:WMS",
            "OGC:WMTS",
            "WMS-C",
        )
    private val DOWNLOAD_SERVICES =
        setOf(
            "OGC:WFS",
            "OGC:WCS",
            "W3C:REST",
            "W3C:WS",
            "W3C:AtomFeed",
        )

    private const val GEONORGE_DOWNLOAD = "GEONORGE:DOWNLOAD"

    fun isViewService(protocol: String?): Boolean = protocol != null && protocol in VIEW_SERVICES

    fun isDownloadService(protocol: String?): Boolean =
        protocol != null && (
            protocol in DOWNLOAD_SERVICES ||
                protocol.startsWith("OGC:API")
        )

    fun appendUuidForGeonorgeDownload(
        url: String,
        protocol: String?,
        uuid: String,
    ): String {
        if (protocol != GEONORGE_DOWNLOAD) return url
        val base = url.trimEnd('/')
        return if (base.endsWith("/$uuid")) base else "$base/$uuid"
    }
}

File: frontend/src/app/metadata/[uuid]/_utils/distributionProtocols.ts
const COPYABLE_DISTRIBUTION_PROTOCOLS = new Set([
  "W3C:REST",
  "OGC:WMS",
  "OGC:WFS",
  "OGC:WCS",
  "OGC:API-Features",
  "OGC:API-Tiles",
  "OPENDAP:OPENDAP",
  "OGC:WMTS",
  "OGC:CSW",
  "OGC:API-Coverages",
  "OGC:OAPIF",
  "W3C:WS",
]);

export const DISTRIBUTION_PROTOCOLS = {
  directDownloadPage: "WWW:DOWNLOAD-1.0-http--download",
  geonorgeDownload: "GEONORGE:DOWNLOAD",
  wms: "OGC:WMS",
} as const;

export function isCopyableDistributionProtocol(protocol: string | null): boolean {
  return protocol != null && COPYABLE_DISTRIBUTION_PROTOCOLS.has(protocol);
}

export function isDirectDownloadPageProtocol(protocol: string | null): boolean {
  return protocol === DISTRIBUTION_PROTOCOLS.directDownloadPage;
}

export function isGeonorgeDownloadProtocol(protocol: string | null): boolean {
  return protocol === DISTRIBUTION_PROTOCOLS.geonorgeDownload;
}

export function isWmsDistributionProtocol(protocol: string | null): boolean {
  return protocol === DISTRIBUTION_PROTOCOLS.wms;
}

